File: src/Mabel.Host.Ios/Sources/MabelHost/Capabilities/CapabilityManifest.swift
import Foundation

// =============================================================================
// Manifesto de capability (lado host) — a declaração capability-based do app.
//
// Espelha src/Mabel.Wasi.Protocol/Capabilities/CapabilityManifest.cs. O host lê
// este JSON (ex.: "mabel.caps.json" no bundle) no load e SÓ liga as capabilities
// declaradas; o resto responde `.notAuthorized` sem tocar o SO.
//
// PLATFORM-AGNOSTIC (Foundation apenas).
// =============================================================================

/// Uma capability concedida, com a justificativa mostrada no prompt do SO e
/// flags opcionais.
public struct CapabilityGrant: Codable, Equatable {
    public let capability: CapabilityId
    public let usageDescription: String?
    public let options: [String: String]?

    public init(capability: CapabilityId, usageDescription: String? = nil,
                options: [String: String]? = nil) {
        self.capability = capability
        self.usageDescription = usageDescription
        self.options = options
    }

    private enum CodingKeys: String, CodingKey {
        case capability, usageDescription, options
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let key = try c.decode(String.self, forKey: .capability)
        guard let cap = CapabilityId(manifestKey: key) else {
            throw DecodingError.dataCorruptedError(
                forKey: .capability, in: c,
                debugDescription: "Capability desconhecida: \(key)")
        }
        self.capability = cap
        self.usageDescription = try c.decodeIfPresent(String.self, forKey: .usageDescription)
        self.options = try c.decodeIfPresent([String: String].self, forKey: .options)
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(capability.manifestKey, forKey: .capability)
        try c.encodeIfPresent(usageDescription, forKey: .usageDescription)
        try c.encodeIfPresent(options, forKey: .options)
    }
}

/// Manifesto do app. Transporte = JSON (mesmo shape do C# / SduiDocument).
public struct CapabilityManifest: Codable, Equatable {
    public let schemaVersion: Int
    public let appId: String
    public let grants: [CapabilityGrant]

    public init(schemaVersion: Int = 1, appId: String, grants: [CapabilityGrant] = []) {
        self.schemaVersion = schemaVersion
        self.appId = appId
        self.grants = grants
    }

    private enum CodingKeys: String, CodingKey {
        case schemaVersion, appId, grants
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.schemaVersion = try c.decodeIfPresent(Int.self, forKey: .schemaVersion) ?? 1
        self.appId = try c.decode(String.self, forKey: .appId)
        self.grants = try c.decodeIfPresent([CapabilityGrant].self, forKey: .grants) ?? []
    }

    /// True se a capability foi declarada (helper de atenuação do host).
    public func isGranted(_ capability: CapabilityId) -> Bool {
        grants.contains { $0.capability == capability }
    }

    public func grant(for capability: CapabilityId) -> CapabilityGrant? {
        grants.first { $0.capability == capability }
    }

    // MARK: - Loading

    /// Decodifica de bytes JSON. Erros propagam (host trata como manifesto inválido).
    public static func decode(from data: Data) throws -> CapabilityManifest {
        try JSONDecoder().decode(CapabilityManifest.self, from: data)
    }

    /// Carrega "mabel.caps.json" do bundle informado. Ausência = manifesto vazio
    /// (app puramente SDUI, zero device access) — decisão fail-closed.
    public static func loadFromBundle(_ bundle: Bundle, name: String = "mabel.caps",
                                      appId: String) -> CapabilityManifest {
        guard let url = bundle.url(forResource: name, withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let manifest = try? decode(from: data)
        else {
            return CapabilityManifest(appId: appId, grants: [])
        }
        return manifest
    }
}

File: src/Mabel.Wasi.Protocol/Capabilities/CapabilityManifest.cs
namespace Mabel.Wasi.Protocol.Capabilities;

// =============================================================================
// Manifesto de capabilities — a declaração capability-based do app.
//
// Princípio (POLA / least authority): o host NÃO liga nenhuma API nativa por
// padrão. Ele lê este manifesto no load e só provê o import real das
// capabilities DECLARADAS. Uma capability não declarada recebe um stub que
// responde CapStatus.NotAuthorized na hora — o guest nunca alcança o SO.
//
// PLATFORM-NEUTRAL: este manifesto é o mesmo pra iOS e Android (alvos co-iguais).
// Cada host — Swift no iOS, Kotlin/Java no Android — lê o mesmo JSON.
//
// Duas camadas de gate (ver ADR 0002 §Segurança):
//   (1) MANIFESTO (host-side, este arquivo): atenuação de autoridade. Estático,
//       auditável, versionado com o app. É o que o host confia.
//   (2) SO/RUNTIME: consentimento do usuário via prompt nativo (iOS e Android
//       runtime permissions). Mesmo declarada, câmera/GPS/notif/biometria ainda
//       precisam do "Permitir".
//
// Ponte com o build (fonte única nas duas plataformas): de CapabilityId +
// UsageDescription, cada host deriva a entrada nativa —
//   iOS:     usage-string no Info.plist (ex.: NSCameraUsageDescription) via xtool;
//   Android: <uses-permission> no AndroidManifest.xml (ex.: android.permission.CAMERA)
//            + a string de rationale do prompt runtime.
// O mapa CapabilityId → permissão nativa é do host. Evita o crash do iOS por
// usage-string ausente e o SecurityException do Android por permissão não
// declarada. Ver docs/capabilities-abi.md §5 (tabelas iOS + Android).
//
// Transporte v1 = JSON (mesmo estilo do SduiDocument). Fica ao lado do app WASM
// no bundle (ex.: "mabel.caps.json").
// =============================================================================

/// <summary>
/// Declaração de UMA capability que o app quer usar, com a justificativa que o
/// SO mostra ao usuário e os metadados de build.
/// </summary>
public sealed record CapabilityGrant
{
    /// <summary>Qual capability. O host liga o import correspondente.</summary>
    public required CapabilityId Capability { get; init; }

    /// <summary>
    /// Texto mostrado no prompt de permissão do SO. No iOS vira a usage-string do
    /// Info.plist (ex.: NSCameraUsageDescription); no Android vira a string de
    /// rationale exibida antes do prompt runtime. Obrigatório para as capabilities
    /// que exigem consentimento; ignorado para as que não exigem
    /// (share/clipboard/haptics).
    /// </summary>
    public string? UsageDescription { get; init; }

    /// <summary>
    /// Flags opcionais por capability (ex.: location "precise" vs "coarse",
    /// notifications "sound"/"badge"). Bag achatado; só o relevante é setado.
    /// </summary>
    public IReadOnlyDictionary<string, string>? Options { get; init; }
}

/// <summary>
/// Manifesto do app. É isto que trafega como JSON e o host carrega no load.
/// </summary>
public sealed record CapabilityManifest
{
    /// <summary>Versão do schema — host recusa/adapta se não reconhecer.</summary>
    public int SchemaVersion { get; init; } = 1;

    /// <summary>Id do app: bundle id no iOS (Info.plist) / applicationId no Android.</summary>
    public required string AppId { get; init; }

    /// <summary>
    /// Capabilities concedidas. O que NÃO estiver aqui é negado por construção
    /// (NotAuthorized). Lista vazia = app puramente SDUI, zero acesso nativo.
    /// </summary>
    public IReadOnlyList<CapabilityGrant> Grants { get; init; } = [];

    /// <summary>True se a capability foi declarada (helper de atenuação do host).</summary>
    public bool IsGranted(CapabilityId capability)
    {
        foreach (var g in Grants)
            if (g.Capability == capability) return true;
        return false;
    }
}
